File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < T; i++) {
            String input = br.readLine();
            if (isValidParenthesisString(input)) {
                sb.append("YES\n");
            } else {
                sb.append("NO\n");
            }
        }

        System.out.print(sb);
    }

    private static boolean isValidParenthesisString(String str) {
        int balance = 0;
        for (char c : str.toCharArray()) {
            if (c == '(') {
                balance++;
            } else if (c == ')') {
                balance--;
            }
            if (balance < 0) { 
                return false;
            }
        }
        return balance == 0; 
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
T = int(input())
for _ in range(T):
    s = input()
    stack = []
    is_vps = True
    for char in s:
        if char == '(':
            stack.append(char)
        else:
            if not stack:
                is_vps = False
                break
            stack.pop()
    if stack:
        is_vps = False
    print("YES" if is_vps else "NO")